### PR TITLE
Actually sum gradient for tf distribute

### PR DESCRIPTION
**Pull request by haohuanw**
https://github.com/keras-team/keras/pull/19467 not actually summing the gradient since it is called in `tf.__internal__.distribute.interim.maybe_merge_call` already. this pr fixes the issue by moving the gradient summing call outside of the merge call.

also updated unit tests to actually checking for the summation.

### CODE REVIEW

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Can you try a numerical comparison with the `tf_keras` package is doing? (which is known to be correct). I recall we tested the current implementation against it and it checked out.

**Comment by haohuanw** [comment]
> Thanks for the PR! Can you try a numerical comparison with the `tf_keras` package is doing? (which is known to be correct). I recall we tested the current implementation against it and it checked out.

added a test in optimizer distributed test to test between keras2 and 3.

File: integration_tests/tf_distribute_training_test.py [hunk 1]
```diff
@@ -50,6 +50,7 @@ def test_model_fit():
             loss=losses.MeanSquaredError(),
             metrics=[metrics.MeanSquaredError()],
         )
+        model.optimizer.build(model.trainable_variables)
```
[the remark this answers is not in the document]
Comment (haohuanw) [reply integration_tests/tf_distribute_training_test.py:53]: it's because optimizer parameters are created lazily - and it is part of `strategy.run` in model which is in per-replica scope. alternatively, i think this can be added some where in `model.fit` or `model.compile`. i actually have to do a similar thing in another test for metrics as well since there are metrics that also creates variables lazily.

[the remark this answers is not in the document]
Comment (fchollet) [reply integration_tests/tf_distribute_training_test.py:53]: But why do the variables need to be created in advance? TensorFlow is supposed to be able to handle lazy variable creation (e.g. at function tracing time). For other backends, we create optimizer and metric variables at the start of `fit()`.

[the remark this answers is not in the document]
Comment (haohuanw) [reply integration_tests/tf_distribute_training_test.py:53]: i think the probably is the variables created inside function being called in `strategy.run(self.one_step_data)` in trainer. in `strategy.run` it overrides the `strategy.scope()` so variable creation was not properly replicated.

[the remark this answers is not in the document]
Comment (fchollet) [reply integration_tests/tf_distribute_training_test.py:53]: We should be fixing that issue first I think -- end users should not be expected to call `build()` themselves. What we could do is start building the optimizer and metrics in `fit()` if a distribution strategy is set. We already do this for the other backends.

[the remark this answers is not in the document]
Comment (haohuanw) [reply integration_tests/tf_distribute_training_test.py:53]: sounds good i will do that in that in this pr.

[the remark this answers is not in the document]
Comment (haohuanw) [reply integration_tests/tf_distribute_training_test.py]: @fchollet should be good now - let me know if this lgty

File: keras/src/callbacks/swap_ema_weights_test.py [hunk 2]
```diff
@@ -40,6 +40,8 @@ def _get_compiled_model(
         model = Sequential(
             [layers.Dense(2, kernel_initializer="ones", use_bias=False)]
         )
+        model.build((None, 3))
+        optimizer.build(model.trainable_variables)
```
Comment (fchollet) [inline keras/src/callbacks/swap_ema_weights_test.py:44]: If the build issue is resolved, you can remove these statements now

File: keras/src/backend/tensorflow/trainer.py [hunk 3]
```diff
@@ -621,6 +631,18 @@ def loss(self, y, y_pred, sample_weight=None):
             x=None, y=y, y_pred=y_pred, sample_weight=sample_weight
         )
 
+    def _symbolic_build(self, iterator=None, data_batch=None):
```
Comment (fchollet) [inline keras/src/backend/tensorflow/trainer.py:634]: I'd like to avoid any undesirable side effects as much as possible, and symbolic build may potentially have some. Please restrict this call to if the distribution strategy is set, so we don't change anything unless symbolic building is actually needed

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

**Comment by haohuanw** [comment]
@fchollet let me know if the new revision lgty

**Review by fchollet — APPROVED** [review]
LGTM, thank you!

**Comment by fchollet** [comment]
Please note, this change appears to break TPU training so we are rolling it back before investigation. https://github.com/keras-team/keras/pull/20017

**Comment by haohuanw** [comment]
@fchollet we can workaround it on our end for now. is there a eta for investigation & follow ups?

**Comment by fchollet** [comment]
Will let you know what we find. We'll seek to have this fix back on master asap.

**Comment by mattdangerw** [comment]
Opened an issue to track. https://github.com/keras-team/keras/issues/20045

Left some details on what was breaking though don't have a simply repro yet. Will work on that.
